drivers/watchdog/wdt_gecko.c

```diff
@@ -0,0 +1,289 @@
+	((struct wdt_gecko_data *)(dev)->driver_data)
+#define DEV_CFG(dev) \
+	((struct wdt_gecko_cfg *)(dev)->config->config_info)
+
+static int wdt_gecko_convert_timeout(u32_t timeout)
+{
+	int idx = 0;
+
+	/* When using ULFRCO (default), 1 cycle is 1 ms +/- 12% */
+	while (idx < ARRAY_SIZE(timeout_in_cycles_tbl)) {
+		if (timeout > timeout_in_cycles_tbl[idx]) {
+			idx++;
+			continue;
+		}
+
+		/* Found the (rounded up?) value of cycles for
+		 *  supplied timeout
+		 */
+		break;
+	}
+
+	return idx;
+}
+
+static int wdt_gecko_convert_window(u32_t window, u32_t period)
+{
+	int idx = 0;
+	u32_t incr_val, comp_val;
+
+	incr_val = period / 8;
+	comp_val = 0; /* Initially 0, disable */
+
+	/* As we're using fixed max window setting of 75% to allow for
+	 * an early warning interrupt, we will only allow a minimum
+	 * window setting of 62.5% (= 5 * 12.5%)
+	 */
+	while (idx < 5) {
+		if (window > comp_val) {
+			comp_val += incr_val;
+			idx++;
+			continue;
+		}
+
+		break;
+	}
+
+	return idx;
+}
+
+static int wdt_gecko_setup(struct device *dev, u8_t options)
+{
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_TypeDef *wdog = config->base;
+
+	if (!data->timeout_valid) {
+		LOG_ERR("No valid timeouts installed");
+		return -EINVAL;
+	}
+
+	data->wdog_config.em2Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+	data->wdog_config.em3Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+
+	data->wdog_config.debugRun =
+		(options & WDT_OPT_PAUSE_HALTED_BY_DBG) == 0U;
+
+	if (data->callback != NULL) {
+		/* Interrupt mode for window */
+		/* Enable timeout and early warning interrupt */
+		WDOGn_IntEnable(wdog, WDOG_IEN_TOUT | WDOG_IF_WARN);
+	} else {
+		/* Disable timeout and early warning interrupt */
+		WDOGn_IntDisable(wdog, WDOG_IEN_TOUT | WDOG_IF_WARN);
+	}
+
+	/* Watchdog is started after initialization */
+	WDOGn_Init(wdog, &data->wdog_config);
+	LOG_DBG("Setup the watchdog");
+
+	return 0;
+}
+
+static int wdt_gecko_disable(struct device *dev)
+{
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_TypeDef *wdog = config->base;
+
+	WDOGn_Enable(wdog, false);
+	data->timeout_valid = false;
+	LOG_DBG("Disabled the watchdog");
+
+	return 0;
+}
+
+static int wdt_gecko_install_timeout(struct device *dev,
+				     const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_Init_TypeDef init_defaults = WDOG_INIT_DEFAULT;
+
+	if (data->timeout_valid) {
+		LOG_ERR("No more timeouts can be installed");
+		return -ENOMEM;
+	}
+
+	if ((cfg->window.max < timeout_in_cycles_tbl[0]) ||
+		(cfg->window.max > timeout_in_cycles_tbl[
+			ARRAY_SIZE(timeout_in_cycles_tbl) - 1])) {
+		LOG_ERR("Upper limit timeout out of range");
+		return -EINVAL;
+	}
+
+	data->wdog_config = init_defaults;
+
+	data->wdog_config.perSel = (WDOG_PeriodSel_TypeDef)
+		wdt_gecko_convert_timeout(cfg->window.max);
+
+	if (cfg->window.min) {
```

Comment:
```
	if (cfg->window.min > 0) {
```
I believe it's not yet a must for the drivers but we're leaning towards writing Misra-C compliant code. At least follow few of the rules that make the code more readable.

Reply:
Thanks for pointing it out, I'll try to stick to it. Changed.